native/src/main/java/io/ballerina/stdlib/serdes/SchemaGenerator.java

```diff
@@ -490,9 +503,15 @@ private static void generateMessageDefinitionForRecordType(ProtobufMessageBuilde
                 case TypeTags.RECORD_TYPE_TAG: {
                     RecordType nestedRecordType = (RecordType) fieldEntryType;
                     String nestedMessageName = nestedRecordType.getName();
-                    ProtobufMessageBuilder nestedMessageBuilder = new ProtobufMessageBuilder(nestedMessageName);
-                    generateMessageDefinitionForRecordType(nestedMessageBuilder, nestedRecordType);
-                    messageBuilder.addNestedMessage(nestedMessageBuilder);
+                    boolean hasMessageDefinition = messageBuilder.hasMessageDefinitionInMessageTree(nestedMessageName);
```

Comment:
It seems like using the state pattern can be used here to simplify the code here, so that maintaining multiple switch-case blocks inside different methods can be avoided. Let's give it a try.

Reply:
+1